### PR TITLE
HoverCard: update native event listeners in case target prop changes.

### LINKED ISSUE #8067: HoverCard: Target props event listener not updating.
On HoverCard, it seems that whenever target props is updated, the mouseEvenetListener for on the target element did not get updated. 

### Environment Information 
 - Release (6.142.1)

### Please provide a reproduction of the bug in a codepen:
https://codepen.io/lauwrentius/pen/GzeyrZ
 
#### Actual behavior:
1. Hover to Element1
2. Click button `Set Hover to Target El2`
3. Hover event triggers when user hover on Element1, while the HoverCard is positioned at Element2
 
#### Expected behavior:
1. Hover to Element1
2. Click button `Set Hover to Target El2`
3. Hover event triggers when user hover on Element2

Are you willing to submit a PR to fix? **Yes**

### PR DESCRIPTION
**Pull request by Vitalius1**
#### Pull request checklist

- [X] Addresses an existing issue: Fixes #8067 
- [X] Include a change request file using `$ npm run change`

#### Description of changes

Reassigning the native event listeners to a new target in case it is updated through props.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/OfficeDev/office-ui-fabric-react/pull/8287)

### CODE REVIEW

**Comment by Jahnp** [comment]
@Vitalius1 , out of curiosity, what's the scenario where this might happen? If say a list is rehydrated -> rerendered and a new target is provided?

**Comment by Vitalius1** [comment]
@Jahnp in most of the cases we know of the usage of HoverCard we actually use it as a wrapper for each target on the page. Whenever that item in the list is rerendered depending on how the rerender happens the target usually is not changed but only the content in it so no listeners reassignment should happen. Also, this is meant to fire only when the consuming app actually keeps a state of the target and passes it down as a prop. In case of large lists, I don't see it practical and this is why we have a HoverCard for each of the elements in the list. In the issue this PR is meant to fix the author actually explains a use case where they reuse the same HoverCard for elements that are replaced by other elements on specific events.

Do you think there is any performance regression with this PR that I might be missing and thus your question?

File: packages/office-ui-fabric-react/src/components/HoverCard/HoverCard.base.tsx [hunk 1]
```diff
@@ -246,4 +234,25 @@ export class HoverCardBase extends BaseComponent<IHoverCardProps, IHoverCardStat
       return prevState;
     });
   };
+
+  private _setEventListeners = (): void => {
+    const { trapFocus, instantOpenOnClick } = this.props;
+    const target = this._getTargetElement();
+    const nativeEventDismiss = this._cardDismiss.bind(this, true);
```
Comment (KevinTCoughlin) [inline packages/office-ui-fabric-react/src/components/HoverCard/HoverCard.base.tsx]: I understand rebinding the `target` event listeners on `target` change, but do we need to rebind this function?

**Comment by KevinTCoughlin** [comment]
> Do you think there is any performance regression with this PR that I might be missing and thus your question?

We're unbinding existing DOM event listeners when the `target` changes, so that's good since it means we won't leak memory by holding onto those DOM nodes which would become "detached".

Binding this many events will have _some_ cost, but it is likely negligible if the action happens sparingly. That's my understanding at least.

**Comment by Vitalius1** [comment]
@KevinTCoughlin @atneik 
Do we even need the binding here... I am confused on why this binding was introduced in the first place?

**Comment by Vitalius1** [comment]
Never mind the previous question... I informed myself by reading the docs for .bind() 😄

**Comment by Vitalius1** [comment]
@Jahnp @atneik I've updated the PR to have the callbacks bind just once. Can you take a look and let me know if there is anything else to address? Thanks!

**Comment by Jahnp** [comment]
Sorry for the delay @Vitalius1 ! My line of thinking was consistent with @KevinTCoughlin 's questions--was just trying to get crisp on the specific scenario you had in mind. Was kind of wondering if there was perhaps a perf usage example/best practice that could be extracted from this, e.g. reusing one HoverCard vs instantiating a new one per-element--that sort of thing.

**Pull request merged**